_gsocblogs/2022/blog_ROOT_JunZhang.md

```diff
@@ -12,24 +13,80 @@ intro: |
+We seem to touch some dark corners in ROOT. During that period, I explored many directions but none seems to work. My mentors and I had some deep discussions, and finally, we came to a new approach: If we cannot ignore those namespaces, why not create a new module that contains all namespaces forward declarations? Because the manually generated module just consists of forward declarations, it's cheap to preload, and ROOT can have all essential knowledge about namespaces with it. So here's the [PR](https://github.com/root-project/root/pull/10910). The performance improvement also looks very promising:
 
-~jun
+![Performance results](https://user-images.githubusercontent.com/77525145/189374093-c6533aa4-8d58-49f1-8996-6605aef0da2b.png)
+
+In short words, we successfully reduced te amount of memory in `CMSSW` by half (from 1.1GB to 600MB) for simple workflows like `hsimple.C`.
+
+I also made some presentations about my work to the team, and they can be found in:
+
+* [Road Map of Optimize ROOT Use of Modules For Large Codebases](https://compiler-research.org/assets/presentations/CaaS_Weekly_08_06_2022_Jun_Zhang_Optimize_ROOT_Use_Of_Modules_For_Large_Codebases.pdf)
+* [Current Working Status of Optimize Modules Usage For Root](https://compiler-research.org/assets/presentations/CaaS_Weekly_07_09_2022_Jun_Zhang_Current_Working_Status_of_Optimize_Modules_Usage_For_ROOT.pdf)
+
+## Post GSoC
+
+However, it's not enough. I have done some deep profiling about Clang, and found out the bottleneck:
+
+```
+% total        Total        Calls  Function
+   99.3  444'432'413       74'169  main [4]
+   92.3  412'998'941       29'027  llvm::SmallVectorBase::grow_pod(void*, unsigned long, unsigned long) [5]
+   89.2  399'211'718        7'870  clang::CompilerInstance::loadModule(clang::SourceLocation, llvm::ArrayRef<std::pair<clang::IdentifierInfo*, clang::SourceLocation> >, clang::Module::NameVisibilityKind, bool) [6]
+   89.2  399'058'838        7'644  clang::ASTReader::ReadAST(llvm::StringRef, clang::serialization::ModuleKind, clang::SourceLocation, unsigned int, llvm::SmallVectorImpl<clang::ASTReader::ImportedSubmodule>*) [7]
+   88.8  397'250'710        8'023  cling::Interpreter::loadModule(clang::Module*, bool) [8]
+   88.7  397'152'678        7'901  clang::Sema::ActOnModuleImport(clang::SourceLocation, clang::SourceLocation, clang::SourceLocation, llvm::ArrayRef<std::pair<clang::IdentifierInfo*, clang::SourceLocation> >) [9]
+   84.0  376'053'039       12'354  TRint::TRint(char const*, int*, char**, void*, int, bool, bool) [10]
+
+```
+
+From the above, we can see the hottest method is `SmallVectorBase::grow_pod`. That is actually caused by a problematic model: Clang always preallocates a huge buffer that is enough to contain all `SourceLocation` in a module, even without use.
+
+To further improve modules in ROOT, I plan to continue the work and cooperate with the upstream, and hopefully, we can solve this longstanding issue and make Clang better.
```

Comment:
We should also mentioned that we pinpointed a real issue in clang and started discussions with the relevant people how to properly solve it rather than trying to hide/mitigate the problem on the side of Cling/ROOT.